# **Exercise 2 Python**

Introduction to Software & Data Engineering with Python - Sekolah Engineering - Pacmann Academy

Materi yang diujikan pada exercise ini adalah
1. Composite Data Structure
2. Looping
3. Function I
4. Function II

# **Composite Data Structure**

### 1

Daftar nilai ujian sebuah kelas adalah sebagai berikut

`nilai = 80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90` 

##### **1A**

Buatlah sebuah list untuk menampung nilai ujian tersebut. Kemudian hitunglah jumlah murid dan carilah nilai tertinggi dan terendah

In [2]:
# buat list yang menampung nilai ujian
nilai=[80, 65, 70, 90, 60, 90, 90, 85, 80, 70, 75, 65, 80, 90, 65, 90]

# Hitung jumlah murid di kelas tersebut
jumlah_murid=len(nilai)

# print jumlah siswa di dalam kelas
print(f'jumlah siswa {jumlah_murid}')

# buat variabel dan print nilai tertinggi
nilai_tertinggi=max(nilai)
print(f"nilai tertinggi {nilai_tertinggi}")

# buat variabel dan print nilai terendah
nilai_terendah=min(nilai)
print(f"nilai terendah {nilai_terendah}")

jumlah siswa 16
nilai tertinggi 90
nilai terendah 60


##### **1B**

Jika siswa memiliki nilai yang kurang dari rata-rata maka siswa harus ujian ulang. Berapakah jumlah siswa yang memiliki nilai dibawah rata-rata


```
Keterangan:
- Bulatkan nilai rata-rata dua nilai dibelakang koma
- gunakan list comprehension untuk nilai yang kurang dari rata-rata
```



In [7]:
# hitunglah dan print nilai rata-rata ujian
avrg=sum(nilai)/len(nilai)
print(f"nilai rata rata adalah {avrg}")

# cari dan print nilai siswa yang kurang dari rata-rata dengan list comprehension
'''nilai_kurang=[]
for i in nilai:
    if i < avrg:
        print(i)
        nilai_kurang.append(i)'''

nilai_kurang=[ i for i in nilai if i < avrg]
print(nilai_kurang)

# hitung dan print jumlah nilai yang kurang dari rata-rata
print(f"nilai kurang ada {len(nilai_kurang)} yaitu {nilai_kurang}")


nilai rata rata adalah 77.8125
[65, 70, 60, 70, 75, 65, 65]
nilai kurang ada 7 yaitu [65, 70, 60, 70, 75, 65, 65]


#### **1C**

Hitunglah prosentase siswa yang lulus dan siswa yang tidak lulus ujian.


In [14]:
# buat variabel dahulu untuk menampung jumlah siswa yang lulus
siswa_lulus=[i for i in nilai if i >= avrg]

# hitung prosentase siswa yang lulus
presentase_lulus=len(siswa_lulus)*100/jumlah_murid

# print prosentase siswa lulus ujian, bulatkan dua angka dibelakang koma
print('presentase siswa yang lulus ujian adalah = %.2f %%' %(presentase_lulus))

# hitung prosentase siswa yang tidak lulus
presentase_tidak_lulus=len(nilai_kurang)*100/jumlah_murid

# print prosentase siswa yang tidak lulus
print('presentase siswa yang tidak lulus ujian adalah = %.2f %%' %(presentase_tidak_lulus))

presentase siswa yang lulus ujian adalah = 56.25 %
presentase siswa yang tidak lulus ujian adalah = 43.75 %


---

### **2**

Penjualan suatu produk pestisida selama tiga bulan dari tiga jenis produk disimpan di dalam data berikut.

|Bulan|Produk 1|Produk 2|Produk 3|
|-|-|-|-|
|Bulan 1|40|45|39|
|Bulan 2|48|50|43|
|Bulan 3|45|48|45|

Angka pada data di atas menunjukkan jumlah produk yang terjual

#### **2A**

Buatlah sebuah tuple of tuple yang menyimpan seluruh data penjualan produk tersebut per bulan. Simpan hanya angkanya saja.


```
Keterangan:
- Struktur tuple
  - my_tuple = ((jumlah_produk_bulan_1, jumlah_produk_bulan_2, jumlah_produk_bulan_3), ...)
```



In [2]:
# definisikan sebuah tuple
my_tuple=((40,45,39),(48,50,43),(45,48,45))

# tampilkan tuple tersebut
print(my_tuple)

((40, 45, 39), (48, 50, 43), (45, 48, 45))


#### **2B**

Carilah rata-rata penjualan setiap bulan dari ketiga produk tersebut.



```
Keterangan:
- bulatkan hasil penjualan menjadi bilangan bulat
```



In [3]:
# gunakan variabel dari soal 2A

#hitung dan tampilkan rata-rata penjualan bulan 1
avg_bulan1=int(sum(my_tuple[0])/len(my_tuple[0]))
print(f"rata2 penjualan bulan1 {avg_bulan1}")

# hitung dan tampilkan rata-rata penjualan bulan 2
avg_bulan2=int(sum(my_tuple[1])/len(my_tuple[1]))
print(f"rata2 penjualan bulan2 {avg_bulan2}")

# hitung dan tampilkan rata-rata penjualan bulan 3
avg_bulan3=int(sum(my_tuple[2])/len(my_tuple[2]))
print(f"rata2 penjualan bulan3 {avg_bulan3}")

rata2 penjualan bulan1 41
rata2 penjualan bulan2 47
rata2 penjualan bulan3 46


#### **2C**

Cari rata-rata penjualan setiap produk selama tiga bulan. Gunakan variabel tuple dari soal 2A.



```
Keterangan:
 - untuk akses tuple of tuple gunakan indeks ganda
 - misalkan ada tuple my_tupe = ((1, 2, 3), (4, 5, 6))
 - misalkan ambil angka 2
    - my_tuple[0][1]
 - indexing pertama untuk mengambil index tuple terluar
 - indexing kedua untuk mengambil nilai yang ada di dalam

```



In [6]:
# Gunakan variabel yang sudah didefinisikan pada soal 2A

# hitung dan print rata-rata penjualan produk 1 selama tiga bulan
rata2_produk1=my_tuple[0][0]+my_tuple[1][0]+my_tuple[2][0]/len(my_tuple)
print(rata2_produk1)

# hitung dan print rata-rata penjualan produk 2 selama tiga bulan
rata2_produk2=my_tuple[0][1]+my_tuple[1][1]+my_tuple[2][1]/len(my_tuple)
print(rata2_produk2)

# hitung dan print rata-rata penjualan produk 3 selama tiga bulan
rata2_produk3=my_tuple[0][2]+my_tuple[1][2]+my_tuple[2][2]/len(my_tuple)
print(rata2_produk3)

103.0
111.0
97.0


### **3**

Lembaga scouting dari sebuah club sepak bola sedang mencari talent muda untuk club mereka. Ada 5 pemain yang masuk incaran, data pemain tersebut sebagai berikut.

|Nama Pemain|Fisik|Mental|Teknik|Menyerang|Bertahan|Posisi|
|-|-|-|-|-|-|-|
|Erling Haaland|89|90|75|95|60|Striker
|Christopher Nkunku|76|78|80|82|65|Sayap Kiri
|Rafael Leao|86|83|85|90|55|Sayap Kiri
|Kevin De Bruyne|82|80|89|88|68|Gelandang Serang
|Vinicius Junior|85|80|85|90|55|Sayap Kiri




#### **3A**

Buatlah sebuah dictionary untuk menyimpan data pemain berikut


In [2]:
# definisikan sebuah dictionary
pemain={'erling haaland':[89,90,75,95,60,'striker'],'Christopher Nkunku':[76,78,80,82,65,'sayap kiri'],'Rafael Leao':[86,83,85,90,55,'Sayap Kiri'],'kevin de bruyne':[82,80,89,88,68,'Gelandang Serang'],'Vinicius Junior':[85,80,85,90,55,'Sayap Kiri']}
# print dictionary tersebut
print(pemain)

{'erling haaland': [89, 90, 75, 95, 60, 'striker'], 'Christopher Nkunku': [76, 78, 80, 82, 65, 'sayap kiri'], 'Rafael Leao': [86, 83, 85, 90, 55, 'Sayap Kiri'], 'kevin de bruyne': [82, 80, 89, 88, 68, 'Gelandang Serang'], 'Vinicius Junior': [85, 80, 85, 90, 55, 'Sayap Kiri']}


#### **3B**

Hitunglah rata-rata kemampuan masing-masing pemain.


```
Keterangan
 - Untuk akses kemampuan pemain gunakan nama_dict[Nama_pemain][kemampuan]
```



In [3]:
# hitung dan print rata-rata kemampuan Erling Haaland
rata2_haland=sum(pemain['erling haaland'][:-1])/len(pemain['erling haaland'][:-1])
print(rata2_haland)
#hitung dan print rata-rata kemampuan Christopher Nkunku
rata2_cris=sum(pemain['Christopher Nkunku'][:-1])/len(pemain['Christopher Nkunku'][:-1])
print(rata2_cris)

# hitung dan print rata-rata kemampuan Rafael Leao
rata2_rafa=sum(pemain['Rafael Leao'][:-1])/len(pemain['Rafael Leao'][:-1])
print(rata2_rafa)

# hitung dan print rata-rata kemampuan Kevin De Bruyne
rata2_kevin=sum(pemain['kevin de bruyne'][:-1])/len(pemain['kevin de bruyne'][:-1])
print(rata2_kevin)

# hitung dan print rata-rata kemampuan Vinicius Junior
rata2_vini=sum(pemain['Vinicius Junior'][:-1])/len(pemain['Vinicius Junior'][:-1])
print(rata2_vini)

81.8
76.2
79.8
81.4
79.0


#### **3C**

Jika dilihat dari rata-rata kemampuan pemain, Christopher Nkunku dan Vinicius Junior merupakan dua yang terendah. Hapus Nkunku dan Vinicius dari dictionary pemain.


```
Keterangan:
 - gunakan method pop()
```



In [4]:
# hapus nkunku dari dictionary pemain
pemain.pop('Christopher Nkunku')


# hapus vinicius dari distionary pemain

pemain.pop('Vinicius Junior')
#  tampilkan dictionary pemain yang baru
pemain

{'erling haaland': [89, 90, 75, 95, 60, 'striker'],
 'Rafael Leao': [86, 83, 85, 90, 55, 'Sayap Kiri'],
 'kevin de bruyne': [82, 80, 89, 88, 68, 'Gelandang Serang']}

#### **3D**

Tersisa tiga pemain yaitu, Erling Haaland, Rafael Leao dan Kevin De Bruyne.
Jika tim scouting sepak bola tersebut sedang mencari penyerang dengan kemampuan **menyerang** terbaik maka siapakah yang akan di ambil ?


```
Keterangan: 
- struktur list
  - [[kemampuan_menyerang, nama_pemain], ...]
- gunakan method max()
```



In [17]:
# buatlah sebuah list dahulu yang menampung kemampuan menyerang tiap pemain
# gunakan variabel dari 3A
penyerang=[['erling haaland',pemain['erling haaland'][3]],['Rafael Leao',pemain['Rafael Leao'][3]],['kevin de bruyne',pemain['kevin de bruyne'][3]]]
best_penyerang=[penyerang[0][1],penyerang[1][1],penyerang[2][1]]

# kemudian cari dan print pemain dengan kemampuan menyerang terbaik
for i in penyerang:
    if i[1]==max(best_penyerang):
        print(f"penyerang terbaik adalah {i[0]} dengan skor {i[1]}")

penyerang terbaik adalah erling haaland dengan skor 95


###**4**

Pak Banu dan Pak Rozi akan bekerja sama dalam berjualan toko buah. Sebelumnya mereka juga sudah mempunyai toko buah, namun karena alasan tertentu mereka memutuskan untuk bekerja sama. Data buah yag di jual Pak Banu dan Pak Rozi adalah sebagai berikut

```
Pak Banu = Jeruk, Apel, Semangka, Rambutan, Pepaya, Kelengkeng, Alpukat, Sirsat
Pak Rozi = Pisang, Melon, Salak, Mangga, Jeruk, Pepaya, Kelengkeng, Srikaya
```





#### **4A**

Buatlah sebuah set untuk menampung buah yang di jual oleh masing-masing penjual

In [43]:
# definisikan sebuah set untuk pak banu
pak_banu={'jeruk','apel','semangka','pepaya','kelengkeng','sirsat'}

# definisikan sebuah set untuk pak rozi
pak_rozi={'pisang','melon','salak','pepaya','kelengkeng','jeruk'}

# print masing-masing set tersebut
print(pak_banu)
print(pak_rozi)

{'semangka', 'kelengkeng', 'pepaya', 'apel', 'sirsat', 'jeruk'}
{'kelengkeng', 'melon', 'salak', 'pisang', 'pepaya', 'jeruk'}


#### **4B**

Gabungkan buah yang dijual Pak Banu dan Pak Rozi


```
Keterangan:
- gunakan method union()
```



In [46]:
# definisikan variabel dan gabungkan buah yang dijual
buah1=pak_banu.union(pak_rozi)
# tampilkan gabungan buah yang dijual
print(buah)

{'semangka', 'kelengkeng', 'melon', 'salak', 'pisang', 'pepaya', 'apel', 'sirsat', 'jeruk'}


#### **4C**

Untuk mempermudah stok ulang buah carilah buah yang sebelumnya sama-sama dijual oleh Pak Banu dan Pak Rozi

```
Keterangan:
- gunakan method intersection()
```

In [47]:
# defisinikan variabel dan cari persamaan buah yang dijual
buah2=pak_banu.intersection(pak_rozi)

# tampilkan buah yang sama antara Pak Banu dan Pak Rozi
print(buah2)

{'pepaya', 'kelengkeng', 'jeruk'}


# **Looping**

### **1**


Pada sebuah puskesmas sedang mengadakan vaksinasi dan memberikan nomor antrian sepanjang 200 orang kepada masyarakat yang ingin vaksin. Untuk **mempercepat** proses vaksinasi puskesmas tersebut **membagi** menjadi empat antrian.

#### **1A**

Untuk nomor antrian kurang dari 100 dibagi berdasarkan bilangan genap dan ganjil. Buatlah urutan bilangan genap dan ganjil menggunakan for loop.

In [19]:
# buat dua empty list untuk menampung deret ganjil dan genap
genap=[i for i in range(0,101) if i%2==0]
ganjil=[i for i in range(0,101) if i%2!=0]

# cari bilangan ganjil dan genap dengan for loop dan if else
# gunakan append untuk menambahkan bilangan ke list
...

# print kedua deret tersebut
print(genap)
print(ganjil)

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100]
[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99]


#### **1B**

Untuk nomor urut 101 sampai 200 dibagi berdasarkan nomor urut
 
 - `Nomor 101 - 125 dan 151 - 175`
 - `Nomor 126 - 150 dan 175 - 200`

Buatlah deret yang menampung dua kondisi tersebut

In [20]:
# definisikan dua list untuk menampung dua kondisi pada soal
urut1=[i for i in range(101,176) if i <=125 or i >=151]
urut2=[i for i in range(126,201) if i <=150 or i >=175]

# cari deret bilangan dengan for loop dan if else
# gunakan append untuk menambahkan bilangan ke list
...
  
# print kedua deret tersebut
print(urut1)
print(urut2)

[101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175]
[126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200]


### **2**

Armada Angkutan Bus Setya Jaya mengeluh karena ada kenaikan harga BBM. Tarif angkutan sebelum kenaikan BBM bisa dilihat pada tabel biaya perjalanan dibawah

|Jurusan | Biaya per Orang | Biaya Pelajar|
|-|-|-|
|Terminal A ke Terminal B | 5.000| 2.500|
|Terminal A ke Terminal C | 10.000 | 4.000|
|Terminal A ke Terminal D | 15.000 | 7.000|


Untuk mengimbangi pengeluaran dan pendapatan para pengusaha angkutan sepakat untuk menaikkan harga.

#### **2A**

Buatlah list yang menampung biaya per orang dan biaya pelajar

In [22]:
# definisikan dua list yang menampung biaya per orang dan biaya pelajar
biaya_per_orang=[5000,10000,15000]
biaya_per_pelajar=[2500,4000,7000]

# print kedua list tersebut
print(biaya_per_orang)
print(biaya_per_pelajar)

[5000, 10000, 15000]
[2500, 4000, 7000]


#### **2B**

Jika **biaya per orang** naik 5.000 rupiah pada semua jurusan terminal, berapa tarif per orang sekarang. Hitung dengan menggunakan **while loop**

In [23]:
# gunakan list yang sudah didefinisikan dari soal 2A

# buat variabel untuk menghentikan while


# hitung tarif baru dengan while
count=0
while count < len(biaya_per_orang):
    biaya_per_orang[count]=biaya_per_orang[count]+5000
    count+=1
# tampilkan list tarif penumpang yang baru
print(biaya_per_orang)

[10000, 15000, 20000]


#### **2C**

Jika tarif penumpang **Pelajar** naik 2.000 rupiah pada semua jurusan terminal, berapa tarif pelajar sekarang ? Hitung menggunakan while loop

In [24]:
# gunakan list yang sudah didefinisikan dari soal 2A

# buat variabel untuk menghentikan while
count=0


# hitung tarif baru dengan while
while count < len(biaya_per_pelajar):
    biaya_per_pelajar[count]=biaya_per_pelajar[count]+2000
    count+=1

# tampilkan list tarif penumpang pelajar yang baru
print(biaya_per_pelajar)

[4500, 6000, 9000]


### **3**

Lembaga scouting dari sebuah club sepak bola sedang mencari talent muda berbakat untuk club mereka. Ada 5 pemain yang masuk incaran, data pemain tersebut sebagai berikut.

|Nama Pemain|Fisik|Mental|Teknik|Menyerang|Bertahan|
|-|-|-|-|-|-|
|Erling Haaland|89|90|75|95|60|
|Christopher Nkunku|76|78|80|82|65|
|Rafael Leao|86|83|85|90|55|
|Kevin De Bruyne|82|80|89|88|68|
|Vinicius Junior|85|80|85|90|55|

#### **3A**

Buatlah dictionary untuk menyimpan data pemain berikut


```
Keterangan:
- Struktur dictionary
  - dict = {nama_pemain:{kemampuan_1:nilai, kemampuan_2:nilai, ...}, ...}
```



In [1]:
# definisikan sebuah dictionary
dict_pemain={'Erling Haaland':{'Fisik':89,'Mental':90,'Teknik':75,'Menyerang':95,'Bertahan':60},'Christopher Nkunku':{'Fisik':76,'Mental':78,'Teknik':80,'Menyerang':82,'Bertahan':65},'Rafael Leao':{'Fisik':86,'Mental':83,'Teknik':85,'Menyerang':90,'Bertahan':55},'Kevin De Bruyne':{'Fisik':82,'Mental':80,'Teknik':89,'Menyerang':88,'Bertahan':68},'Vinicius Junior':{'Fisik':85,'Mental':80,'Teknik':85,'Menyerang':90,'Bertahan':55}}
# print dictionary tersebut
print(dict_pemain)

{'Erling Haaland': {'Fisik': 89, 'Mental': 90, 'Teknik': 75, 'Menyerang': 95, 'Bertahan': 60}, 'Christopher Nkunku': {'Fisik': 76, 'Mental': 78, 'Teknik': 80, 'Menyerang': 82, 'Bertahan': 65}, 'Rafael Leao': {'Fisik': 86, 'Mental': 83, 'Teknik': 85, 'Menyerang': 90, 'Bertahan': 55}, 'Kevin De Bruyne': {'Fisik': 82, 'Mental': 80, 'Teknik': 89, 'Menyerang': 88, 'Bertahan': 68}, 'Vinicius Junior': {'Fisik': 85, 'Mental': 80, 'Teknik': 85, 'Menyerang': 90, 'Bertahan': 55}}


#### **3B**

Club membutuhkan pemain bola dengan mental yang bagus. Jika ada pemain dengan **mental** lebih kecil dari **80** keluarkan dari data scouting.

In [9]:
# untuk menghindari eror pada saat loop, gunakan method copy untuk menyalin dictionary pada variabel yang baru
# buat nama variabel dahulu kemudian copy dictionary yang sudah dibuat sebelumnya
mental_pemain=dict_pemain.copy()
# gunakan for dan if else untuk menyeleksi pemain
# jika pemain masuk memenuhi syarat pakai "continue"
# jika tidak memenuhi syarat, langsung keluarkan dari dict, pakai method pop()
for i in dict_pemain.keys():
    if dict_pemain[i]['Mental']>=80:
        continue
    else:
        mental_pemain.pop(i)



# gunakan list comprehension untuk menampilkan pemain yang tersisa
# buat variabel baru dahulu untuk menyimpan hasil list comprehension
lst_mental_pemain=[i for i in mental_pemain.keys()]
print(lst_mental_pemain)


['Erling Haaland', 'Rafael Leao', 'Kevin De Bruyne', 'Vinicius Junior']


#### **3C**


Pengamat Sepak Bola menyarankan untuk langsung mengambil pemain dengan kemampuan **menyerang** lebih dari **92**. Namun pihak scouting masih mempertimbangkan hal lain. Untuk itu pihak scouting hanya akan melihat adakah pemain dengan kemampuan lebih dari **92**.

In [12]:
# gunakan cara yang sama agar saat loop dictionary tidak eror yaitu menyalin dictionary ke variabel baru
# bedakan nama variabel copy dengan yang sebelumnya
penyerang_pemain=dict_pemain.copy()

# pakai for loop dan if else untuk mencari pemain dengan kemampuan menyerang lebih dari 92
# pakai pass jika menemukan pemain dengan kondisi tersebut
for i in dict_pemain.keys():
    if dict_pemain[i]['Menyerang']>=92:
        continue
    else:
        penyerang_pemain.pop(i)

lst_penyerang_pemain=[i for i in penyerang_pemain.keys()]
print(f" Pemain dengan kemampuan menyerang lebih dari sama dengan 92 : {lst_penyerang_pemain[0]}")
    

 Pemain dengan kemampuan menyerang lebih dari sama dengan 92 : Erling Haaland


#### **3D**

Pelatih membutuhkan pemain dengan **teknik** lebih dari **85** dan **fisik** lebih dari sama dengan **84**.
Jika ada pemain yang tidak memenuhi syarat tersebut langsung keluarkan dari data scouting.

In [17]:
# gunakan cara yang sama agar saat loop dictionary tidak eror yaitu menyalin dictionary ke variabel baru
# bedakan nama variabel copy dengan yang lain
teknik_fisik=dict_pemain.copy()


# gunakan for dan if else untuk menyeleksi pemain
# jika pemain masuk memenuhi syarat pakai "continue"
# jika tidak memenuhi syarat, langsung keluarkan dari dict, pakai method pop()
for i in dict_pemain.keys():
    if dict_pemain[i]['Teknik']>=85 and dict_pemain[i]['Fisik']>=84:
        continue
    else:
        teknik_fisik.pop(i)

lst_teknik_fisik=[i for i in teknik_fisik.keys()]
print(teknik_fisik)

{'Rafael Leao': {'Fisik': 86, 'Mental': 83, 'Teknik': 85, 'Menyerang': 90, 'Bertahan': 55}, 'Vinicius Junior': {'Fisik': 85, 'Mental': 80, 'Teknik': 85, 'Menyerang': 90, 'Bertahan': 55}}


#### **3E**

Masih tersisa dua pemain yaitu Rafael Leao dan Vinicius Junior. Uang kas klub hanya bisa digunakan untuk membeli satu pemain saja, untuk itu cari pemain dengan rata-rata kemampuan paling tinggi dari kedua pemain tersebut

In [25]:
# gunakan cara yang sama agar saat loop dictionary tidak eror yaitu menyalin dictionary ke variabel baru
# bedakan nama variabel copy dengan yang lain
rata2_pemain=teknik_fisik.copy()

# definisikan variabel untuk menampung rata-rata
rata2_pemain2={}

# gunakan for untuk menyeleksi pemain
# didalam for yang pertama, definisikan variabel jumlah untuk menampung jumlah total kemampuan pemain
# anda bisa memakai method keys atau cara lain pada for yang kedua jika anda pakai nested-for
for i in teknik_fisik:
    average= sum(teknik_fisik[i].values())/len(teknik_fisik[i])
    print(f"rata-rata kemampuan {i} adalah {average}")

rata-rata kemampuan Rafael Leao adalah 79.8
rata-rata kemampuan Vinicius Junior adalah 79.0


### **4**

Bu Surti akan berbelanja di swalayan dengan Bu Mima dan Bu Rani


```
belanjaan ibu surti = 20.000, 80.000, 40.000, 150.500
belanjaan ibu mima = 31.000, 43.000, 65.000, 105.000
belanjaan ibu rani = 12.000, 200.000, 61.000, 10.000
```

#### **4A**

Buatlah list untuk menampung harga belanja masing-masing ibu tersebut

In [26]:
# buat tiga list untuk menampung list harga belanjaan
surti=[20_000,80_000,40_000,150_500]
mima=[31_000,43_000,65_000,105_000]
rani=[12_000,200_000,61_000,10_000]

#### **4B**

Cari total dari belanjaan masing-masing ibu dengan list comprehension, jika ada harga yang lebih dari 50.000 diskon 5%

In [39]:
# Cari total belanja dengan list comprehension
# HINT, Gunakan method sum()
total_belanja_surti=sum([i-(i*0.05) if i > 50_000 else i for i in surti])
total_belanja_mima=sum([i-(i*0.05) if i > 50_000 else i for i in mima])
total_belanja_rani=sum([i-(i*0.05) if i > 50_000 else i for i in rani])
# tampilkan masing-masing total belanjaan
print(f"Total belanja bu Surti = Rp {total_belanja_surti}")
print(f"Total belanja bu Mima = Rp {total_belanja_mima}")
print(f"Total belanja bu Rani = Rp {total_belanja_rani}")

Total belanja bu Surti = Rp 278975.0
Total belanja bu Mima = Rp 235500.0
Total belanja bu Rani = Rp 269950.0


# **Function I**

---
## Case 1 : Waktu Mencuci Tangan
Dianjurkan bagi seseorang untuk mencuci tangan sampai 21 detik secara menyeluruh agar tangan dapat dikatakan bersih dari kuman dan kotoran. 

Buatlah fungsi yang menerima input jumlah seseorang mencuci tangan dalam satu hari dan lama waktu orang tersebut mengikuti anjuran cuci tangan (dalam bulan). 

Hitung banyaknya waktu (dalam menit dan detik) yang dihabiskan orang tersebut untuk mencuci tangan.   

- anggap 1 bulan == 30 hari

```
Contoh:
Setelah mengetahui anjuran tersebut, selama 2 bulan terakhir Athaya mencuci tangan sesuai dengan anjuran sebanyak 6 kali dalam sehari

input  
6 kali dan 2 bulan

ouput :
Waktu total mencuci tangan adalah 126 menit dan 0 detik

```

### 1A.
Buatlah fungsi tersebut!
```
Catatan tambahan:
  - Gunakan paramater yang dapat menerima input jumlah cuci tangan dan jumlah bulan
  - Gunakan return pada fungsi untuk mengembalikan output yang didapat
```

In [ ]:
#membuat fungsi
def mencuci_tangan(a,b):
    hari=a*21
    total=hari*30*b
    return



###1B.
Panggil fungsi tersebut! Apabila,

a. ) Selama 1 tahun 1 bulan terakhir Richard mencuci tangan sesuai dengan anjuran sebanyak 3 kali dalam sehari

In [ ]:
#memanggil fungsi 
...

'Waktu total mencuci tangan adalah : 409 menit dan 30 detik'

b. ) Selama 3 bulan terakhir, Nabila mencuci tangan sesuai anjaran sebanyak 5 kali dalam sehari

In [ ]:
#memanggil fungsi 
...

'Waktu total mencuci tangan adalah : 157 menit dan 30 detik'

---
##Case 2 : Perkalian
Buatlah fungsi yang menghasilkan list hasil perkalian! 

Sebagai contoh jika user meng-inputkan angka 7. Berarti fungsi akan menampilkan perkalian 7 x 1 sampai 7 x 10.

```
Contoh:
input  
7

ouput :
[7, 14, 21, 28, 35, 42, 49, 56, 63, 70]
```


###2A.
Membuat fungsi **tabel perkalian**.
```
Catatan:
 - Parameter yang digunakan adalah angka dan panjang tabel perkalian
 - Karena panjang tabel perkalian hingga 10. Maka buatlah default value = 10
 - Gunakan return untuk mengembalikan nilai
 - Anda dapat menggunakan looping untuk mengembangkan fungsi
```

In [ ]:
#membuat fungsi tabel_perkalian
...

###2B.
Memanggil fungsi **tabel_perkalian** dengan:

a.) Tampilkan tabel perkalian 7

In [ ]:
#memanggil fungsi
...

[7, 14, 21, 28, 35, 42, 49, 56, 63, 70]

b.) Tampilkan tabel perkalian 121

In [ ]:
#memanggil fungsi
...

[121, 242, 363, 484, 605, 726, 847, 968, 1089, 1210]

c.) Kemudian, anda ingin menampilkan tabel perkalian 10 tetapi dengan panjang tabel hingga 15

Tampilkan nilai tersebut tanpa merubah fungsi
```
Catatan:
  - Anda dapat menggunakan keyword arguments
```

In [ ]:
#memanggil fungsi
...

[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150]

---
##Case 3 : Perhitungan BMI
Anda bekerja disuatu pekerjaan dibidang kesehatan dan diminta untuk membuat kalkulator BMI (Body Mass index) sederhana yang dapat menentukan apakah orang tersebut mengalami berat badan dibawah normal, berat badan normal, kelebihan berat badan, atau obesitas berdasarkan berat badan dan tinggi badan yang dimasukkan. Fungsi dikembangkan dengan ketentuan sebagai berikut :

###3A.
Buatlah sebuah fungsi **bmi** yang dapat menampung input dari user yang berisi berapa berat badan user (dalam kilogram) dan berapa tinggi badan user (dalam sentimeter)

```
🔽 Ketika fungsi dijalankan, akan menghasilkan 🔽
 Masukkan berat badan (dalam kg) :
 Masukkan tinggi badan (dalam cm) : 
 
```

In [ ]:
#buat fungsi
...

Untuk mengecek apakah fungsi yang anda buat dapat berjalan sesuai kriteria atau tidak, anda dapat memanggil fungsi bmi

In [ ]:
#memanggil fungsi
...

 Masukkan berat badan (dalam kg) : 50
 Masukkan tinggi badan (dalam cm) :  150


###3B.
Kembangkanlah fungsi yang sudah dibuat agar dapat menghitung BMI dari input yang sudah diberikan dan menentukan kategori BMI.

| BMI |Kategori|
|-|-|
|**<18,5**  | Berat Badan dibawah normal| 
|**18,5 - 22,9**  | Berat badan normal | 
|**23 - 24,9**  | Kelebihan berat badan | 
|**>= 25**  | Obesitas | 

```
Keterangan :
 - Input berat badan dan tinggi badan dari user yang akan digunakan untuk melakukan perhitungan BMI
 - Rumus perhitungan BMI = berat badan(kg) : ( tinggi badan(meter) x tinggi badan(meter) )
 - Bulatkan hasil BMI hingga 2 angka dibelakang koma.
 - Anda dapat memanfaatkan branching untuk mengembangkan fungsi.
 - Untuk meminimalisir terjadinya error, ketika user input selain angka / number. Maka program akan menampilkan "Input harus berupa angka” 

🔽Contoh output ketika fungsi dijalankan, akan menghasilkan 🔽
Masukkan berat badan (dalam kg) : 60
Masukkan tinggi badan (dalam cm) :  162

Body Mass Index (BMI) : 22.86
Kategori : Berat normal
```

In [ ]:
#Mengembangkan fungsi agar dapat menghitung dan menetukan BMI
...

Untuk dapat mengecek apakah fungsi yang anda buat sesuai dengan kriteria diminta atau tidak, anda dapat melakukan pemanggilan fungsi **bmi**

input A :
Berat badan Ara = 60 kg
Tinggi Badan Ara = 162 cm

input B :
Berat badan Ahmad = 62 kg
Tinggi Badan Ahmad = 187,4 cm

In [ ]:
#memanggil fungsi dengan input A
...

Masukkan berat badan (dalam kg) : 60
Masukkan tinggi badan (dalam cm) :  162

Body Mass Index (BMI) : 22.86
Kategori : Berat normal


In [ ]:
#memanggil fungsi dengan input B
...

Masukkan berat badan (dalam kg) : 62
Masukkan tinggi badan (dalam cm) :  187.4

Body Mass Index (BMI) : 17.65
Kategori : Berat badan dibawah normal


---
##Case 4
Sebuah sekolah sedang melakukan survey mengenai pengeluaran transportasi (berangkat-pulang sekolah) perbulan yang dikeluarkan siswa SD A selama sebulan 

Untuk menunjang survey pengeluaran, anda diminta untuk membuat fungsi dengan menggunakan arbitrary keyword arguments yang dapat menampilkan pasangan nama siswa dan besaran biaya transport (rupiah) yang dikeluarkan perbulannya.


###4A. 
Membuat fungsi menggunakan arbitrary keyword arguments
```
Catatan:
 - Nilai yang dimasukkan adalah nama siswa dan biaya pengeluaran transportasi perbulannya

🔽 Ketika fungsi dipanggil, output yang diharapkan 🔽
  Pengeluaran Niriina Amalina untuk transportasi senilai : Rp.150000
```

In [ ]:
#membuat fungsi
...

###4B.
Beberapa hari kemudian, telah didapat sampel sementara dari survey yang dijabarkan dalam tabel berikut:

| Nama Siswa |Pengeluaran Transportasi Perbulan|
|-|-|
|Randy | Rp. 240.000 | 
|Alfin| Rp. 200.000 | 
|Shafira | Rp. 175.000 | 
|Andi  | Rp. 320.000| 
|Nuning  | Rp. 400.000 | 

Cobalah panggil fungsi berdasarkan data diatas!

In [ ]:
#memanggil fungsi
...

Pengeluaran Randy untuk transportasi senilai : Rp.240000
Pengeluaran Alfin untuk transportasi senilai : Rp.200000
Pengeluaran Shafira untuk transportasi senilai : Rp.175000
Pengeluaran Andi untuk transportasi senilai : Rp.320000
Pengeluaran Nuning untuk transportasi senilai : Rp.400000


---
##Case 5 : Docstring Perhitungan Faktorial
Terdapat fungsi **print_faktor** yang dapat mengetahui faktorisasi dari suatu angka:

In [ ]:
def print_faktor(angka):
    print("faktor dari ",angka,"adalah:")
    for i in range(1, angka + 1):
        if angka % i == 0:
            print(i, end = ' ')

angka_input = 50

In [ ]:
print_faktor(angka_input)

faktor dari  50 adalah:
1 2 5 10 25 50 

Akan tetapi, fungsi tersebut masih sulit untuk dipahami. Maka dari itu, buatlah docstring dari fungsi diatas agar mempermudah user dalam memahami program!

In [ ]:
#membuat docstring
#copy paste fungsi diatas dan tambahkan docstring
...


##Case 6 : Perhitungan Pendapatan Perkapita
Anda ingin mengetahui tingkat kemakmuran negara dan masyarakat melalui perhitungan Pendapatan PerKapita (PPK). 

Maka anda berencana untuk membuat fungsi **ppk** yang dapat menghitung hasil pendapatan rata-rata masyarakat pada suatu negara

```
Catatan:
 Rumus pendapatan perkapita = pendapatan nasional : jumlah penduduk
```

###6A.
Buatlah fungsi tersebut!

```
Catatan :     
 - Didalam fungsi, argumen yang akan diinputkan adalah nama negara, pendapatan nasional suatu negara, dan total jumlah penduduk.

🔽Contoh output ketika fungsi dijalankan, akan menghasilkan 🔽
Pendapatan negara x adalah y milyar rupiah perpenduduk.
 ```

In [ ]:
#membuat fungsi ppk
...

Panggil fungsi tersebut dengan menggunakan data dibawah ini

| input ke -1 | Nama Negara |Pendapatan Nasional (Miliar Rupiah) | Jumlah Penduduk (Juta Jiwa)
| - |-|-| -|
|1. |Indonesia | 210 | 200 |  
|2. |Malaysia | 75 | 25 |
|3. |Vietnam | 80 | 67 | 

In [ ]:
# memanggil fungsi dengan input ke-1
...

'Pendapatan negara Indonesia adalah 1.05 milyar rupiah perpenduduk.'

In [ ]:
# memanggil fungsi dengan input ke-2
...

'Pendapatan negara Malaysia adalah 3.0 milyar rupiah perpenduduk.'

In [ ]:
# memanggil fungsi dengan input ke-3
...

'Pendapatan negara Vietnam adalah 1.1940298507462686 milyar rupiah perpenduduk.'

###6B.
Kemudian, untuk memudahkan dalam memahami isi dari fungsi, anda berencana untuk membuat docstring fungsi ppk

In [ ]:
#fungsi ppk
#membuat docstring dari fungsi ppk
...

###6C.
Anda berencana untuk membuat fungsi baru ppk dengan menggunakan fungsi lambda agar fungsi lebih ringkas dan sederhana.
```
🔽Contoh output ketika fungsi dijalankan, akan menghasilkan 🔽
Pendapatan negara x adalah y milyar rupiah perpenduduk.
 ```

In [ ]:
#membuat fungsi ppk dengan lambda
...

###6D.
Panggil fungsi lambda dari data Indonesia!

In [ ]:
# memanggil fungsi 
...

'Pendapatan negara Indonesia adalah 1.05 milyar rupiah perpenduduk.'

# **Function II**

---
##Case 1 : Kebutuhan Kalori Harian
Sebuah lembaga kesehatan berencana untuk melakukan penelitian mengenai pengaruh kebutuhan kalori / Basal Metabolic Rate (BMR) pada pekerja kantor yang aktivitas sehari-harinya duduk seharian penuh didepan komputer.

Untuk pengembangan proyek penelitian, anda diminta untuk membuat modul python berisi fungsi-fungsi menghitung kebutuhan kalori harian pria dan wanita.

###1A.
Untuk langkah awal, buatlah modul **bmr** yang berisi fungsi perhitungan kebutuhan kalori untuk pria dan wanita.

Kemudian, buatlah fungsi baru **pria** yang berisi perhitungan kebutuhan kalori berdasarkan rumus Harris Beneddict.
```
Catatan :
 - Argumen yang akan diinputkan adalah berat badan (kg), tinggi badan (cm), dan usia(tahun)
 - Rumus menghitung BMR pria = ( 88,362 + (13,397 x berat badan) + (4,799 x tinggi badan) – (5,677 x usia) ) x 1,2
 - Bulatkan hasil BMR dua angka dibelakang koma

🔽 Output ketika program dijalankan, akan menghasilkan 🔽
Kebutuhan kalori yang dibutuhkan sebesar :


```

In [ ]:
#modul bmr
#fungsi pria
...

###1B.
Selanjutnya, buatlah fungsi baru **wanita** yang berisi perhitungan kebutuhan kalori berdasarkan rumus Harris Beneddict.
```
Catatan :
 - Argumen yang akan diinputkan adalah berat badan (kg), tinggi badan (cm), dan usia(tahun)
 - Rumus menghitung BMR wanita = ( 447,593 + (9,247 x berat badan) + (3,098 x tinggi badan) – (4,330 x usia) ) x 1,2
 - Bulatkan hasil BMR dua angka dibelakang koma

🔽 Output ketika program dijalankan, akan menghasilkan 🔽
Kebutuhan kalori yang dibutuhkan sebesar :


```

In [ ]:
#modul bmr
# fungsi pria
...

# fungsi wanita
...

###1C.
Setelah modul disimpan, anda ingin mengecek apakah modul yang anda bangun dapat berjalan atau tidak.

Untuk itu anda berencana mencoba membuat file utama untuk mengecek apakah modul dapat berjalan dengan baik. Impor modul **bmr** yang telah dibuat sebelumnya.



In [ ]:
#impor modul
...

###1D.
Coba panggil fungsi **pria** dengan nilai input:
  - Berat badan Wahyu = 70 kg
  - Tinggi Badan Wahyu = 181 kg
  - Usia Wahyu = 25 tahun



In [ ]:
#akses fungsi pria
...

Kebutuhan kalori yang dibutuhkan sebesar : 


2103.42

###1E.
Coba panggil fungsi **wanita** dengan nilai input:
  - Berat badan Rita = 62 kg
  - Tinggi Badan Rita = 158 kg
  - Usia Rita  = 27 tahun



In [ ]:
#akses fungsi wanita
...  

Kebutuhan kalori yang dibutuhkan sebesar : 


1672.18

---
##Case 2 : Laporan Penjualan
Terdapat laporan penjualan perusahaan Z di Bulan April yang disajikan dalam bentuk tabel dibawah.

|Laporan Penjualan Bulan April | Keterangan |
|-|-|
|Penjualan | Rp. 100.000.000|
|Potongan Penjualan |Rp. 15.000.000|
|Retur Pembelian | Rp. 2.000.000|
|Persediaan Barang |Rp. 75.000.000|
|Persediaan Akhir |Rp. 15.000.000|


###2A.
Buatlah sebuat file python baru dengan nama **main.py** dan tambahkan dictionary berisi tabel Laporan Penjualan Bulan April.

In [ ]:
#file main.py
#dictionary
...

In [ ]:
#coba cek isi dictionary, apakah sudah sesuai dengan data atau belum
...

dict_items([('penjualan', 100000000), ('potongan_penjualan', 15000000), ('retur', 2000000), ('persediaan_barang', 75000000), ('persediaan_akhir', 1500000)])

###2B.
Selanjutnya, anda diminta untuk membuat modul baru dengan nama **laporan penjualan** yang berisi perhitungan harga pokok penjualan(HPP), penjualan bersih, dan laba kotor dari tabel diatas,

Buatlah **fungsi hpp** dari modul laporan penjualan
```
Catatan :
 - Rumus menghitung harga pokok penjualan = persediaan barang - persediaan akhir 
 - Value diambil langsung dari dictionary yang dibuat sebelumnya di main file.
```

In [ ]:
#modul laporan_penjualan
#fungsi hpp
...

###2C.
Kemudian buatlah **fungsi penjualan bersih** dari modul laporan penjualan
```
Catatan :
 - Rumus menghitung penjualan bersih = penjualan - (potongan penjualan + retur pembelian) 
 - Value diambil langsung dari dictionary yang dibuat sebelumnya di main file.
```

In [ ]:
# modul laporan_penjualan
# fungsi hpp
...

# fungsi penjualan_bersih
...

###2D.
Kemudian buatlah **fungsi laba kotor** dari modul laporan penjualan
```
Catatan :
 - Rumus menghitung laba kotor = penjualan bersih - harga pokok penjualan
 - Value diambil langsung dari dictionary yang dibuat sebelumnya di main file.
```

In [ ]:
# modul laporan_penjualan
# fungsi hpp
...

# fungsi penjualan_bersih
...

# fungsi laba kotor
...

###2E.
Simpanlah modul python tersebut. Kemudian, untuk mencoba apakah modul dapat berjalan dengan baik di **main.py**. Import dan jalankan fungsi-fungsi yang ada didalamnya.

Import modul


In [ ]:
# main.py
# import modul
...

# dictionary yang telah dibuat sebelumnya
...

###2F.
Panggil fungsi **hpp** dengan nilai input dari dictionary yang telah dibuat sebelumnya.



In [ ]:
# main.py
# import modul
...

# dictionary yang telah dibuat sebelumnya
...
#memanggil fungsi hpp
...

73500000

###2G.
Setelah nilai harga pokok penjualan telah didapat. Simpan nilai hpp kedalam dictionary agar nantinya dapat digunakan kembali oleh program.




In [ ]:
# main.py
# import modul
...

# dictionary yang telah dibuat sebelumnya
...

#memanggil fungsi hpp
...

#simpan nilai hpp kedalam dictionary
...

###2H.
Panggil fungsi **penjualan_bersih** dengan nilai input dari dictionary yang telah dibuat sebelumnya.



In [ ]:
#memanggil fungsi penjualan bersih
...

83000000